source/fundamentals/utf8-validation.txt

```diff
@@ -0,0 +1,116 @@
+.. _nodejs-utf-8-validation:
+
+================
+UTF-8 Validation
+================
+
+.. default-domain:: mongodb
+
+.. contents:: On this page
+   :local:
+   :backlinks: none
+   :depth: 2
+   :class: singlecol
+
+Overview
+--------
+
+In this guide, you can learn how to enable or disable the {+driver-short+}'s
+**UTF-8** validation feature. **UTF-8** is a character encoding specification
+that ensures compatibility and consistent presentation across most operating
+systems, applications, and language character sets. By default, the driver
+checks documents for any characters that are not encoded in a valid UTF-8
+format when it transfers data between your application and MongoDB.
+
+If you enable validation, the driver substitutes invalid UTF-8 characters
+with valid UTF-8 ones when transferring data between your application and
+MongoDB. The validation adds processing overhead since it needs to verify the
+data and to make substitutions when necessary.
+
+If you disable validation, your application can avoid the processing
+overhead caused by validation, but cannot guarantee consistent presentation of
+invalid UTF-8 data.
+
+Read the sections below to learn how to set UTF-8 validation using the
+{+driver-short+}.
+
+.. note::
+
+   The {+driver-short+} automatically substitutes invalid UTF-8 characters
+   in your application before sending them to MongoDB regardless of whether
+   you enable validation. You can only retrieve invalid UTF-8 characters
+   from MongoDB.
```

Comment:
[question] Are invalid UTF-8 characters that are stored in the database inadvertently substituted when read/written back, even with this setting set to false? 

1. with UTF validation set to false, retrieve invalid UTF-8 characters from MongoDB as part of a string field
2. make a change to that string that does not affect the invalid character
3. update the database with the new string, still containing the invalid character

Is this what you mean by "cannot guarantee consistent presentation of invalid UTF-8 data."? I don't think anything needs to change, but for me this note immediately raises the question "what if I want to keep my existing invalid UTF characters safely stored in MongoDB?"

Reply:
> Are invalid UTF-8 characters that are stored in the database inadvertently substituted when read/written back, even with this setting set to false?

They're converted in your application by the js-bson library as mentioned here: "The Node.js driver automatically substitutes invalid UTF-8 characters in your application before sending them to MongoDB regardless of whether you enable validation." You can't access any actual invalid UTF-8 from the driver, so if you try to insert what you fetched from MongoDB, what gets written is not the same value, but rather a replacement).

Was this admonition unclear, and if so, what would you suggest?

Reply:
OK that makes sense. "You can only retrieve invalid UTF-8 characters from MongoDB" is very clear. 

What's less clear to me is whether you can work with invalid UTF-8 characters in your application after reading them from MongoDB. 

Substituting invalid characters "before sending them to MongoDB" makes it sound like the substitution is part of a write operation and you could work with invalid characters in your application up until you write them back to MongoDB. If that's correct I think the note is clear and nothing needs to be changed - I just wanted to be sure I understood that you might read an invalid character and then write back a substitution.

---

One way to clarify that all of this only applies to read operations might be in the **Specify the UTF-8 Validation Setting** section - it mentions calling CRUD operations (CRUD is mentioned a few more times after), should this just be Read operations? All the examples show reads. That may or may not work as a clarification; I don't know if you can supply this to create/update/delete.

Reply:
> What's less clear to me is whether you can work with invalid UTF-8 characters in your application after reading them from MongoDB if you do not enable validation.

From what I read in the code, I think that you can, but it depends in what capacity. You won't be able to write that data back to MongoDB as invalid UTF-8, but you have access to it in your application. I don't know whether it's in scope to talk about potential use cases for this.


> Substituting invalid characters "before sending them to MongoDB" makes it sound like the substitution is part of a write operation and you could work with invalid characters in your application up until you write them

I've reworded this, but it is true that you can work with invalid characters in your application up until you write them unless you're already working with BSON (not sure that's an option in the Node.js driver). When the driver calls the js-bson library to serialize the data to BSON, it performs the utf8 conversion, so any invalid UTF-8 is substituted by an alternative UTF-8 value.

> One way to clarify that the validation setting only applies to read operations might be in the Specify the UTF-8 Validation Setting section - it mentions calling CRUD operations (CRUD is mentioned a few more times after), should this just be Read operations?

I don't know whether it's true that validation only applies to read operations, but you can provide it to any CRUD operation. I think there might be a non-read operation case in which `update()` returns the updated document in a callback, and that could contain invalid UTF-8 in one of the fields that wasn't updated. Maybe the DBX engineers can verify this (I haven't dug deeper into storing invalid UTF-8 in MongoDB since it seemed out of scope for this page).

Reply:
Got it, thanks for explaining :+1: